### PR TITLE
InstructorFeedbackQuestionEditAction: Condition of if-statement is always true #4723

### LINKED ISSUE #4723: InstructorFeedbackQuestionEditAction: Condition of if-statement is always true
In `InstructorFeedbackQuestionEditAction.java` around line 84, it seems that the condition of the if-statement is always true even if the question number is not edited.

### PR DESCRIPTION
**Pull request by unyoungwax**
Fixes #4723

### CODE REVIEW

**Comment by kanghj** [comment]
@unyoungwax, assign a reviewer for your pull request from now on.

File: src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java [hunk 1]
```diff
@@ -93,9 +96,13 @@ private void testGeneralQuestionOperations() throws Exception {
         testNewQuestionLink();
         testInputValidationForQuestion();
         testAddQuestionAction();
+        
+        testResponseRateBeforeEditingQuestion();
 
         testEditQuestionLink();
         testEditQuestionAction();
+        
+        testResponseRateAfterEditingQuestion();
```
Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java]: It might be better to not test it this way. Instead, will it be better to include the tests for response rate to be in <code>testEditQuestionAction</code>?

File: src/test/resources/pages/instructorFeedbackQuestionEditToTeamToTeam.html [hunk 2]
```diff
@@ -738,11 +738,8 @@ <h5 class="col-xs-2" data-original-title="Select this option to use a custom tim
          </form>
       </div>
       <br/>
-      <div class="overflow-auto alert alert-success" id="statusMessage">
-         The question has been added to this feedback session.
+      <div id="statusMessage" style="display: none;">
```
Comment (kanghj) [inline src/test/resources/pages/instructorFeedbackQuestionEditToTeamToTeam.html]: This may weaken this particular test. It will be better to craft a test for this issue that doesn't weaken the other existing tests.

File: src/main/java/teammates/logic/backdoor/BackDoorLogic.java [hunk 3]
```diff
@@ -152,9 +152,9 @@ public String persistDataBundle(DataBundle dataBundle)
             response = injectRealIds(response);
         }
         frDb.createFeedbackResponses(responses.values());
-        
-        for(FeedbackSessionAttributes session : sessions.values()){
-            updateRespondants(session.feedbackSessionName, session.courseId);
+
+        for (FeedbackResponseAttributes response : responses.values()) {
+            updateRespondants(response.feedbackSessionName, response.courseId);
```
Comment (kanghj) [inline src/main/java/teammates/logic/backdoor/BackDoorLogic.java]: Remove the modifications to this file, if you can, since in the end nothing was changed here (other than indentation). Logically, none of your commits should touch this file.

Comment (unyoungwax) [inline src/main/java/teammates/logic/backdoor/BackDoorLogic.java]: The new backdoor operation that I added (`createFeedbackResponse`) takes in `FeedbackResponseAttributes` as the only parameter. I modified this file so that `BackDoorLogic` requires only `FeedbackResponseAttributes` to create a feedback response (previously `BackDoorLogic` requires both `FeedbackSessionAttributes` and `FeedbackResponseAttributes`).

File: src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java (hunk 1 above)
Comment (unyoungwax) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java]: I will combine `testResponseRateBeforeEditingQuestion` and `testResponseRateAfterEditingQuestion` into one method. The method should neither affect existing tests nor depend on them.

**Comment by unyoungwax** [comment]
I have created a new test method `testResponseRate` and reverted changes made to the existing tests.

File: src/main/java/teammates/logic/backdoor/BackDoorLogic.java (hunk 3 above)
Comment (kanghj) [inline src/main/java/teammates/logic/backdoor/BackDoorLogic.java]: Oops my bad, didn't notice that somehow

File: src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java [hunk 4]
```diff
@@ -80,46 +80,30 @@ private void editQuestion(FeedbackQuestionAttributes updatedQuestion) throws Inv
...
+
+        // if error is not empty not tested as extractFeedbackQuestionData
+        // method uses Assumptions to cover it
+        if (!questionDetailsErrors.isEmpty()) {
+            statusToUser.addAll(questionDetailsErrorsMessages);
+            isError = true;
         } else {
-            List<String> questionDetailsErrors = updatedQuestion.getQuestionDetails().validateQuestionDetails();
-            List<StatusMessage> questionDetailsErrorsMessages = new ArrayList<StatusMessage>();
+            logic.updateFeedbackQuestionNumber(updatedQuestion);
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java]: Is the old method, <code>updateFeedbackQuestion</code> still used in the code?

File: src/test/java/teammates/test/driver/BackDoor.java [hunk 5]
```diff
@@ -555,6 +555,13 @@ public static String deleteFeedbackQuestion(String questionId) {
     private void ____FEEDBACK_RESPONSE_level_methods______________________________() {
     }
     
+    public static String createFeedbackResponse(FeedbackResponseAttributes feedbackResponse) {
+        DataBundle dataBundle = new DataBundle();
+        dataBundle.feedbackResponses.put("dummy-key", feedbackResponse);
+        return persistNewDataBundle(Utils.getTeammatesGson()
+                .toJson(dataBundle));
```
Comment (kanghj) [inline src/test/java/teammates/test/driver/BackDoor.java]: indentation is a bit off here

File: src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java [hunk 6]
```diff
@@ -447,6 +452,53 @@ private void testEditNonExistentQuestion() {
...
+                        Const.DEFAULT_SECTION, 
+                        "alice.b.tmms@gmail.tmt",
+                        Const.DEFAULT_SECTION, 
+                        new Text("Response from instructor to Alice"));
+        BackDoor.createFeedbackResponse(feedbackResponse);
+
+        ______TS("check response rate before editing question");
+        
+        AppUrl feedbacksPageUrl = createUrl(Const.ActionURIs.INSTRUCTOR_FEEDBACKS_PAGE).withInstructorId(instructorId);
+        InstructorFeedbacksPage feedbacksPage = AppPage.getNewPageInstance(browser, feedbacksPageUrl, InstructorFeedbacksPage.class);
```
Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java]: Extract this into a new method. Also, check if using <code>AppPage#changePageType</code> is more suitable here.

File: src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java [hunk 7]
```diff
@@ -447,6 +452,53 @@ private void testEditNonExistentQuestion() {
...
+        feedbacksPage.waitForPageToLoad();
+        
+        assertEquals("1 / 1", feedbacksPage.getResponseValue(courseId, feedbackSessionName));
```
Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java]: While this is fine too, I am wondering if we can do this test without navigating to the feedbacks page. Perhaps this test would be more suitable as an action level test..

File: src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java (hunk 4 above)
Comment (unyoungwax) [inline src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java]: No, I have changed `updateFeedbackQuestion` to `updateFeedbackQuestionWithResponseRateCheck` in `updateFeedbackQuestionNumber` method.

File: src/test/java/teammates/test/driver/BackDoor.java (hunk 5 above)
Comment (unyoungwax) [inline src/test/java/teammates/test/driver/BackDoor.java]: I follow the style of similar methods, e.g. `createAccount`.

File: src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java (hunk 7 above)
Comment (unyoungwax) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java]: I will create an action level test. Should I keep the browser test?

Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java]: I think we can still keep the browser test for now, since you have already written it, unless it's clear that it has no value after the action test has been added.

File: src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java (hunk 4 above)
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java]: Hmm, I meant the entire codebase. If <code>updateFeedbackQuestion</code> is not used anywhere in the code, we can remove it and rename <code>updateFeedbackQuestionNumber</code> to updateFeedbackQuestion. Will reduce code duplication, since currently updateFeedbackQuestionNumber and updateFeedbackQuestion are very similar anyway. Could be done in a separate issue if it's too big a change though.

Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java]: Oh wait, I got confused again. Ignore that

File: src/test/java/teammates/test/cases/ui/InstructorFeedbackQuestionEditActionTest.java [hunk 8]
```diff
@@ -1202,6 +1202,73 @@ public void testExecuteAndPostProcessRubricQuestion() throws Exception {
...
+        removeAndRestoreTypicalDataInDatastore();
+        
+        gaeSimulation.loginAsInstructor(dataBundle.instructors.get("instructor1OfCourse1").googleId);
+
+        FeedbackSessionsLogic fsLogic = FeedbackSessionsLogic.inst();
+        FeedbackQuestionsLogic fqLogic = FeedbackQuestionsLogic.inst();
+
+        FeedbackSessionAttributes fs = dataBundle.feedbackSessions.get("session1InCourse1");
+
+        // Check response rate before editing question 1
```
Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/InstructorFeedbackQuestionEditActionTest.java]: For splitting up logical parts of the test, use <code>______TS</code> to be consistent with the rest of the code

File: src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java [hunk 9]
```diff
@@ -80,46 +80,30 @@ private void editQuestion(FeedbackQuestionAttributes updatedQuestion) throws Inv
...
-            // if error is not empty not tested as extractFeedbackQuestionData method uses Assumptions to cover it
-            if (!questionDetailsErrors.isEmpty()) {
-                statusToUser.addAll(questionDetailsErrorsMessages);
-                isError = true;
-            } else {
-                logic.updateFeedbackQuestionNumber(updatedQuestion);
-                statusToUser.add(new StatusMessage(Const.StatusMessages.FEEDBACK_QUESTION_EDITED, StatusMessageColor.SUCCESS));
-            }
+
+        List<String> questionDetailsErrors = updatedQuestion.getQuestionDetails().validateQuestionDetails();
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java]: This is not really your part of the code, but could you make a variable for <code>updatedQuestion.getQuestionDetails()</code>, since it seems to be used again below?

File: src/test/java/teammates/test/cases/ui/InstructorFeedbackQuestionEditActionTest.java [hunk 10]
```diff
@@ -1202,6 +1202,73 @@ public void testExecuteAndPostProcessRubricQuestion() throws Exception {
...
+        fs = fsLogic.getFeedbackSession(fs.feedbackSessionName, fs.courseId);
+        assertEquals(4, fsLogic.getFeedbackSessionDetails(fs).stats.submittedTotal);
```
Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/InstructorFeedbackQuestionEditActionTest.java]: I think we can check for both <code>submittedTotal</code> and <code>expectedTotal</code>

File: src/test/java/teammates/test/cases/ui/InstructorFeedbackQuestionEditActionTest.java [hunk 11]
```diff
@@ -1202,6 +1203,106 @@ public void testExecuteAndPostProcessRubricQuestion() throws Exception {
...
+
+        FeedbackSessionsLogic fsLogic = FeedbackSessionsLogic.inst();
+        FeedbackQuestionsLogic fqLogic = FeedbackQuestionsLogic.inst();
+
+        FeedbackSessionAttributes fs = dataBundle.feedbackSessions.get("session1InCourse1");
+
+        ______TS("Check response rate before editing question 1");
+        fs = fsLogic.getFeedbackSession(fs.feedbackSessionName, fs.courseId);
+        FeedbackSessionDetailsBundle details = fsLogic.getFeedbackSessionDetails(fs);
+        assertEquals(4, details.stats.submittedTotal);
```
Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/InstructorFeedbackQuestionEditActionTest.java]: Maybe extract these repeated numbers into local variables in this class

File: src/test/java/teammates/test/cases/ui/InstructorFeedbackQuestionEditActionTest.java [hunk 12]
```diff
@@ -1202,6 +1203,106 @@ public void testExecuteAndPostProcessRubricQuestion() throws Exception {
...
+        a = getAction(params3);
+        a.executeAndPostProcess();
+
+        // Response rate should decrease by 1 because question 3 has a unique respondent
+        fs = fsLogic.getFeedbackSession(fs.feedbackSessionName, fs.courseId);
+        details = fsLogic.getFeedbackSessionDetails(fs);
+        assertEquals(3, details.stats.submittedTotal);
+        assertEquals(10, details.stats.expectedTotal);
+
+        ______TS("Change the feedback path of question 4");
```
Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/InstructorFeedbackQuestionEditActionTest.java]: @unyoungwax, Hmm, instead of calling it "question 4", it will be better to provide some context of what this test is testing.

File: src/test/java/teammates/test/cases/ui/InstructorFeedbackQuestionEditActionTest.java [hunk 13]
```diff
@@ -1202,6 +1203,106 @@ public void testExecuteAndPostProcessRubricQuestion() throws Exception {
...
+        InstructorFeedbackQuestionEditAction a = getAction(params1);
+        a.executeAndPostProcess();
+
+        // Response rate should not change because other questions have the same respondents
+        fs = fsLogic.getFeedbackSession(fs.feedbackSessionName, fs.courseId);
+        details = fsLogic.getFeedbackSessionDetails(fs);
+        assertEquals(4, details.stats.submittedTotal);
+        assertEquals(10, details.stats.expectedTotal);
+
+        ______TS("Change the feedback path of question 3");
```
Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/InstructorFeedbackQuestionEditActionTest.java]: @unyoungwax, likewise here, give a better context than just calling it "question 3"

**Comment by unyoungwax** [comment]
I have improved the comments and removed the magic numbers in the new test case.

File: src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java [hunk 14]
```diff
@@ -566,7 +566,7 @@ public void updateFeedbackQuestionNumber(FeedbackQuestionAttributes newQuestion)
         }
         
         adjustQuestionNumbers(oldQuestionNumber, newQuestionNumber, questions);
-        updateFeedbackQuestion(newQuestion);
+        updateFeedbackQuestionWithResponseRateCheck(newQuestion);
```
Comment (kanghj) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: Actually, I have a suspicion that the old <code>updateFeedbackQuestion</code> should not be used anywhere in the entire code at all. Can you check that for me? We can make a new issue and fix that there if so.

File: src/test/java/teammates/test/cases/ui/InstructorFeedbackQuestionEditActionTest.java [hunk 15]
```diff
@@ -1202,6 +1203,117 @@ public void testExecuteAndPostProcessRubricQuestion() throws Exception {
...
+        a = getAction(params3);
+        a.executeAndPostProcess();
+
+        // Response rate should decrease by 1 because the response of the unique instructor respondent is deleted
+        fs = fsLogic.getFeedbackSession(fs.feedbackSessionName, fs.courseId);
+        details = fsLogic.getFeedbackSessionDetails(fs);
+        assertEquals(numStudentRespondents, details.stats.submittedTotal);
+        assertEquals(totalStudents + totalInstructors, details.stats.expectedTotal);
+
+        ______TS("Change the feedback path of a question so that some possible respondents are removed");
```
Comment (kanghj) [inline src/test/java/teammates/test/cases/ui/InstructorFeedbackQuestionEditActionTest.java]: Actually I feel a bit disturbed by these test, because now it will harder to add or modify existing questions in the typical data bundle, because we make these assumptions about the questions and responses here. I guess it's still fine though, since we won't expect much changes to the typical data bundle, but this is something to think about in future when you add tests.

File: src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java (hunk 14 above)
Comment (unyoungwax) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: The old `updateFeedbackQuestion` is used in `adjustQuestionNumbers`, `shiftQuestionNumbersDown`(only the question numbers are changed) and  `FeedbackQuestionLogicTest` (can be changed to `updateFeedbackQuestionWithResponseRateCheck`). If the old `updateFeedbackQuestion` is removed, the response rate will be checked again and again (no changes made) when adjusting the question numbers.

Comment (kanghj) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: Oops, I updated my old comment. I meant to decrease the visibility of the old method from <code>public</code> to <code>private</code>, so in future developers can't accidentally use that method instead.

Comment (unyoungwax) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: I have the same idea too, the old `updateFeedbackQuestion` should be made private. The comment should also mention that the response rate is not updated.

Comment (kanghj) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: Ok, let's do that in another issue then, since it seems it will affect multiple places in the code

Comment (damithc) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: @unyoungwax Please create a new issue for that if you haven't done so already. In fact you can go ahead and take the issue.

File: src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java [hunk 16]
```diff
@@ -80,46 +80,31 @@ private void editQuestion(FeedbackQuestionAttributes updatedQuestion) throws Inv
...
+        FeedbackQuestionDetails updatedQuestionDetails = updatedQuestion.getQuestionDetails();
+        List<String> questionDetailsErrors = updatedQuestionDetails.validateQuestionDetails();
+        List<StatusMessage> questionDetailsErrorsMessages = new ArrayList<StatusMessage>();
+
+        for (String error : questionDetailsErrors) {
+            questionDetailsErrorsMessages.add(new StatusMessage(error, StatusMessageColor.DANGER));
+        }
+
+        // if error is not empty not tested as extractFeedbackQuestionData
+        // method uses Assumptions to cover it
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/InstructorFeedbackQuestionEditAction.java]: This comment is a bit hard to understand.

File: src/test/java/teammates/test/driver/BackDoor.java (hunk 5 above)
Comment (damithc) [inline src/test/java/teammates/test/driver/BackDoor.java]: No need for the line wrapping at all? 
8 space indentation is the last resort when line wrapping. There are better alternatives. It's ok to use the best indentation, even if other methods don't use it.

**Comment by damithc** [comment]
Some minor cosmetic changes requested.

**Comment by unyoungwax** [comment]
I have removed the outdated comment in `InstructorFeedbackQuestionEditAction` and also used the correct indentation for the code in `BackDoor`.

**Comment by kanghj** [comment]
@unyoungwax, remember to merge this

**Comment by kanghj** [comment]
@unyoungwax, the merge commit's message is wrong, take note of that next time. You have to merge locally on your computer and run all tests before merging to master

**Comment by unyoungwax** [comment]
I apologize for that. I have run the tests locally before merging.

File: src/main/java/teammates/logic/backdoor/BackDoorLogic.java (hunk 3 above)
Comment (wkurniawan07) [inline src/main/java/teammates/logic/backdoor/BackDoorLogic.java]: So this seems to be the guilty change. @unyoungwax I don't seem to get the reason for this. Why is this needed?
